bluesky_kafka/tools/queue_thread.py

```diff
@@ -0,0 +1,213 @@
+        (name, document) tuples placed on this queue will be published as Kafka messages
+        by kafka_publisher; by default a queue.Queue will be used
+    publisher_queue_timeout: float (optional)
+        time in seconds to wait for a document to become available on the publisher_queue
+        before checking if the publisher thread should terminate; default is 1s
+
+    Returns
+    -------
+    publisher_queue: queue.Queue-like object
+        (name, document) tuples placed on this queue will be published as Kafka messages
+        by kafka_publisher. If no Kafka broker can be found the returned publisher_queue
+        will be None.
+    publisher_thread_stop_event: threading.Event
+        call set() on this threading.Event to terminate the message publication loop;
+        this may have unexpected memory consequences if (name, document) tuples continue
+        to be placed on publisher_queue
+    """
+
+    logger = logging.getLogger("bluesky_kafka")
+
+    # let's not confuse the publisher_queue parameter, which is None by default,
+    # with the queue that will be returned, which is None only in a failure mode
+    final_publisher_queue = None
+    publisher_thread_stop_event = None
+
+    try:
+        logger.info("connecting to Kafka broker(s): '%s'", bootstrap_servers)
+        # verify the specified topic exists on the Kafka broker(s) before subscribing
+        topic_to_topic_metadata = list_topics(bootstrap_servers=bootstrap_servers)
+        if topic in topic_to_topic_metadata:
+            # since the topic exists, build a Publisher for the topic
+            kafka_publisher = Publisher(
+                topic=topic,
+                bootstrap_servers=bootstrap_servers,
+                # specify a key to guarantee messages will be delivered in order
+                key=str(uuid.uuid4()),
+                producer_config=producer_config,
+                flush_on_stop_doc=True,
+            )
+            (
+                final_publisher_queue,
+                publisher_thread,
+                publisher_thread_stop_event,
+            ) = start_kafka_publisher_thread(
+                publisher_queue=publisher_queue,
+                publisher=kafka_publisher,
+                publisher_queue_timeout=publisher_queue_timeout,
+            )
+            logger.info(
+                "RunEngine will publish bluesky documents on Kafka topic '%s'",
+                topic,
+            )
+        else:
+            raise BlueskyKafkaException(
+                f"topic `{topic}` does not exist on Kafka broker(s) `{bootstrap_servers}`",
+            )
+    except BaseException:
+        """
+        An exception at this point means Kafka
+        messages will not be published.
+        """
+        logger.exception(
+            "Kafka messages can not be published on topic '%s'",
+            topic,
+        )
+
+    return final_publisher_queue, publisher_thread_stop_event
```

Comment:
maybe return the thread object as well?

Reply:
maybe return a namedtuple or dataclass here?  Particularly if you also return a callback that can be subscribed to the RE directly.